src/psbt/mod.rs

```diff
@@ -894,117 +897,137 @@ impl PsbtInputExt for psbt::Input {
-                                if tapleaf_hashes.last() != Some(&tapleaf_hash) {
-                                    tapleaf_hashes.push(tapleaf_hash);
-                                }
-                            })
-                            .or_insert_with(|| {
-                                (
-                                    vec![tapleaf_hash],
-                                    (xpk.master_fingerprint(), xpk.full_derivation_path()),
-                                )
-                            });
-                    }
+                        })
+                        .or_insert_with(|| {
+                            (
+                                vec![tapleaf_hash],
+                                (xpk.master_fingerprint(), xpk.full_derivation_path()),
+                            )
+                        });
                 }
             }
+        }
 
-            derived
-        } else {
-            let bip32_derivation = RefCell::new(BTreeMap::new());
-            let derived = descriptor.translate_pk2(|xpk| {
-                let derived = xpk.derive_public_key(&secp)?;
-                bip32_derivation.borrow_mut().insert(
-                    derived.to_public_key().inner,
-                    (xpk.master_fingerprint(), xpk.full_derivation_path()),
-                );
-                Ok(derived)
-            })?;
-
-            self.bip32_derivation = bip32_derivation.into_inner();
-
-            match &derived {
-                Descriptor::Bare(_) | Descriptor::Pkh(_) | Descriptor::Wpkh(_) => {}
-                Descriptor::Sh(sh) => match sh.as_inner() {
-                    descriptor::ShInner::Wsh(wsh) => {
-                        self.witness_script = Some(wsh.inner_script());
-                        self.redeem_script = Some(wsh.inner_script().to_v0_p2wsh());
-                    }
-                    descriptor::ShInner::Wpkh(..) => self.redeem_script = Some(sh.inner_script()),
-                    descriptor::ShInner::SortedMulti(_) | descriptor::ShInner::Ms(_) => {
-                        self.redeem_script = Some(sh.inner_script())
-                    }
-                },
-                Descriptor::Wsh(wsh) => self.witness_script = Some(wsh.inner_script()),
-                Descriptor::Tr(_) => unreachable!("Tr is dealt with separately"),
+        derived
+    } else {
+        let bip32_derivation = RefCell::new(BTreeMap::new());
+        let derived = descriptor.translate_pk2(|xpk| {
+            let derived = xpk.derive_public_key(&secp)?;
+            bip32_derivation.borrow_mut().insert(
+                derived.to_public_key().inner,
+                (xpk.master_fingerprint(), xpk.full_derivation_path()),
+            );
+            Ok(derived)
+        })?;
+
+        if let Some(check_script) = check_script {
+            if check_script != derived.script_pubkey() {
+                return Ok((derived, false));
```

Comment:
I opted to not do `Option<...>` because it would mean an `unwrap` for one of the unchecked caller which seemed like the worse option. I added a comment explaining the awkwardness.